### ISSUE TITLE
CLI: CLI Event Streaming Tool for local sandbox

### ISSUE THREAD

**Issue by paulbellamy** [stellar/soroban-tools#39]
Idea is a local CLI tool to stream events for devs, such as something like:
```
$ cli events --contract-id 01
{"type": "Transfer", "data": {...}}
```

Relevant CAP: https://github.com/stellar/stellar-protocol/issues/1280
Soroban-RPC API: https://docs.google.com/document/d/1TZUDgo_3zPz7TiPMMHVW_mtogjLyPL0plvzGMsxSz6A/edit#heading=h.t03hji406t76

### Background & Planning work

- [ ] Design a storage format and file for events in the local sandbox
  - `.soroban/events.json` with a list of events serialized to json might be fine. or xdr?
  - The storage and transfer mechanism here should support linux/mac/windows
- [ ] Add the command to the CLI
  - `soroban-cli events`
  - Default behaviour is to print the latest N messages from the sandbox. (or all messages? see what kubectl logs does).
  - Figure out a nice human-friendly format to print the events in.
- [ ] stream events from the sandbox as they come in.
  - `soroban-cli events -f`
- [ ] add `--id CONTRACT_ID` flag for per-contract filtering (or `--id` to match other functions?)
- [ ] Add `getEvents` method support in `soroban serve`

## Nice-to-haves

- [ ] allow output in json (strval) format, as well as xdr?

**Comment by paulbellamy** [comment]
Blocked on https://github.com/stellar/stellar-core/issues/3487

**Comment by paulbellamy** [comment]
Deferring this to sept/october, as I don't think it is critical for the first e2e example

**Comment by tomerweller** [comment]
> HTTP api for this in cli http server (or move to new issue)

+1 on this. 

This almost seems more important than streaming to the standard output. Standard output is great for debug but, if the developer wants to use the events, wouldn't they prefer an HTTP endpoint? 

More so, assuming that RPC servers use the same API spec, this should provide a smooth transition from local testing to production RPC, no?

**Comment by leighmcculloch** [comment]
I think this functionality is important for developer testing their applications in sandbox, but it should be a small lift in addition to the HTTP endpoint since they'd both be backed by the same functionality.

**Comment by leighmcculloch** [comment]
> $ cli events --contract-id 01

This subcommand structure makes sense I think for a deployed contract, but less so for a contract running in sandbox. For a contract running in the sandbox and being invoked via `soroban-cli invoke`, I think the events should just be dumped to stdout as part of the output of the invoke command.

So I think this issue probably needs splitting into two issues:
1. `soroban-cli events` streaming events for deployed contracts.
2. `soroban-cli invoke` streaming events during invoke.

**Comment by paulbellamy** [comment]
See: https://github.com/stellar/rs-soroban-env/pull/342

**Comment by leighmcculloch** [comment]
@paulbellamy I took a look at #97 but it's unclear what we're building this for. In sandbox mode there are two ways contracts are invoked, via the invoke command and via the pretend RPC server. How does the events command fit into that? As I mentioned above, when I'm invoking a contract via the command line I'd expect to see events as log output immediately during the invoke.

I expected the event streaming command would be for deployed contracts on futurenet. Is that what this is building for? If this is for sandbox, how is it to work?

**Comment by paulbellamy** [comment]
There is a bit of a multiples here. You can trigger events via `cli invoke` or via `cli serve`'s `sendTransaction` method. And you can consume them via `cli serve`'s `eventsSubscribe` method, or via the proposed `cli events`.


The consuming via `cli serve` is the more important part here, imo. That lets you develop your events processing against the sandbox. I'd expect you could connect your consumer to the `cli serve` endpoint and trigger events (either via `cli invoke` or via `cli serve`s `sendTransaction` method) and get the events in the consumer.

I agree that printing the events directly for the user (maybe to stderr?) would be fine for `cli invoke`). In which case maybe we skip the `cli events` subcommand for now. It really just adds symmetry to the usage, imo.

**Comment by leighmcculloch** [comment]
I'm missing some context and so the above isn't clear to me.

Is the idea that `cli events` is a client to `cli serve`, and that client streams events from the serve and outputs them to stdout?

**Comment by paulbellamy** [comment]
No, it would be similar to `cli invoke`

**Comment by smac711** [comment]
@leighmcculloch @paulbellamy I see this issue still in the august iteration should this be completed by the end of this weeks sprint or should we have this part of the plan for the next iteration starting Monday? 

Is this required for Meridian?

**Comment by leighmcculloch** [comment]
> No, it would be similar to `cli invoke`

In what ways would it be similar to invoke? I'm failing to get a complete picture of how cli events will be used. If I run `cli events` what is it connecting to to get events, and where is it sending the events it collects?

> printing the events directly for the user (maybe to stderr?) would be fine for cli invoke

I've opened https://github.com/stellar/soroban-cli/issues/128, because afaict we don't have an issue for that yet.

**Comment by 2opremio** [comment]
@leighmcculloch I am a little confused. We already seem to be printing the events at the end of the invocation (https://github.com/stellar/soroban-cli/pull/105 ) . If we don't include `events` subcommand wouldn't this issue be a NOP?

**Comment by leighmcculloch** [comment]
@2opremio I'm pretty confused about what the `events` subcommand does when coupled with `invoke`. I thought the use case for `events` was only:

> soroban-cli events streaming events for deployed contracts.

-- @leighmcculloch Ref: https://github.com/stellar/soroban-cli/issues/39#issuecomment-1218320836

But it seems it also has use with invoke, but I'm failing understand it.

**Comment by paulbellamy** [comment]
The only part of this that is important for meridian is the http part of it. I’m tired of debating the `cli events` part, it is a completion/polish priority. If we have to split the http part into a different issue so it can get done, then let's do that.

**Comment by tomerweller** [comment]
> The only part of this that is critical for meridian is the http part of it

Agreed. 

> If we have to split the http part into a different issue so it can get done, then let's do that.

Also agreed. Let's break this issue to two separate ones : one for the HTTP portion (critical/pre-meridian) and one for `cli events` (polish/backlog)?

**Comment by mollykarcher** [comment]
Closing as we've since killed off the sandbox
